### PR TITLE
Ds 3138

### PR DESCRIPTION
**Pull request by JustinCCYap**

### CODE REVIEW

File: R/firstlast.R [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+# The number of time periods in units of 'by' between the start and end dates
+# as a integer (rounded down)
+#' @importFrom lubridate interval time_length
+intervalLength <- function(start.date.time, end.date.time, by)
+{
+    if (by == "quarter")
+        floor(intervalLength(start.date.time, end.date.time, "month") / 3)
+    else
+        floor(time_length(interval(start.date.time, end.date.time), by))
+}
```
Comment (jrwishart) [inline R/firstlast.R:247]: Could unit test these internal functions, `intervalLength` and `periodIntegers` and `parseDateTime`?

File: DESCRIPTION [hunk 2]
```diff
@@ -1,16 +1,24 @@
 Package: verbs
 Type: Package
 Title: A Grammar for Common Data Manipulations and Summaries
-Version: 0.0.1
+Version: 0.0.2
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply DESCRIPTION:4]: Done by @jrwishart

File: DESCRIPTION [hunk 3]
```diff
@@ -1,16 +1,24 @@
...
 Author: Displayr <opensource@displayr.com>
 Maintainer: Displayr <opensource@displayr.com>
 Description: A collection of operations/actions/"verbs" providing an intuitive interface for many common operations (e.g. Sum, First, Average, Maximum, ...) on various inputs/data structures. Created to simplify common user actions in Displayr with a wide variety of inputs. 
 Suggests:
     testthat    
 License: GPL-3
+Imports:
+    flipTime,
+    flipU,
+    lubridate
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply DESCRIPTION:14]: I use `interval` and `time_length` from lubridate

File: R/firstlast.R [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    else
+        paste0("The duration '", by,
+               "' cannot be applied as dimension ", dimension.index,
+               " in the input data is ")
+
+    if (is.null(date.time.strings))
+        stop(common.msg, "not labeled with dates.")
+    date.times <- AsDateTime(date.time.strings, on.parse.failure = "")
+    if (any(is.na(date.times)))
+        stop(common.msg, "labeled with invalid date(s).")
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:214]: It is not possible at the moment to determine the invalid dates with AsDateTime. It just returns a vector of NA.

File: R/firstlast.R [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+#'   between start and end dates).
+#' @param ... Additional arguments for \code{head} and \code{tail}.
+#' @return A subset containing the last elements in x
+#' @examples
+#'   Last(1:10, 6) # 4:10
+#'   x <- 1:10
+#'   names(x) <- Sys.Date() - 1:10
+#'   Last(x, keep = 1, by = "week", calendar = FALSE) # previous 7 days
+#' @importFrom utils tail
+#' @export
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:62]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    if (!is.atomic(keep) || is.array(keep) || !allWholeNumbers(keep))
+        stop("The input 'keep' needs to be an integer scalar or vector ",
+             "containing the number of entries to keep corresponding to ",
+             "the dimensions of the input data.")
+
+    # Check lengths of x and 'keep'
+    if ((is.null(dim.x) && length(keep) > 1))
+        stop("The input 'keep' is a vector with more than one value. It needs to ",
+             "be scalar when 'x' is non-dimensioned, e.g., when it is a ",
+             "vector.")
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:88]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: tests/testthat/test-firstlast.R [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    # vector, matches head and tail
+    x <- 1:10
+    names(x) <- LETTERS[1:10]
+    expect_equal(First(x), head(x))
+    expect_equal(Last(x, 2), tail(x, 2))
+
+    # matrix, matches head and tail
+    m <- matrix(1:20, 4)
+    expect_equal(First(m), head(m))
+    expect_equal(Last(m, 2), tail(m, 2))
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply tests/testthat/test-firstlast.R:22]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: tests/testthat/test-firstlast.R [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+    # matrix with multiple keep, matches head and tail
+    m <- matrix(1:20, 4)
+    expect_equal(First(m, c(2, 3)), head(m, c(2, 3)))
+    expect_equal(Last(m, c(NA, 1)), tail(m, c(NA, 1)))
+
+    # array, matches head and tail
+    arr <- array(1:24, dim = c(2,3,4))
+    expect_equal(First(arr), head(arr))
+    expect_equal(Last(arr, 2), tail(arr, 2))
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply tests/testthat/test-firstlast.R:32]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    is.numeric(v) && all(is.na(v) | (round(v) == v & v >= 0))
+}
+
+# Is valid duration, ignoring NA
+validDuration <- function(x)
+{
+    pattern <- paste0("^[[:blank:]]*[[:digit:]]*[[:blank:]]*(",
+                      paste0(allowed.time.units, collapse = "|"),
+                      ")s*[[:blank:]]*$")
+    is.character(x) && all(is.na(x) | grepl(pattern, x, ignore.case = TRUE))
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:151]: Removed in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+             "be scalar when 'x' is non-dimensioned, e.g., when it is a ",
+             "vector.")
+    if ((is.null(dim.x) && length(keep) > 1) ||
+        (!is.null(dim.x) && length(keep) > length(dim.x)))
+        stop("The input 'keep' is a vector with length greater than the number of ",
+             "dimensions of 'x'. Its length needs to be less than or equal to ",
+             "the number of dimensions of 'x'.")
+
+    if (length(keep) > 1 && all(is.na(keep)))
+        stop("The input 'keep' cannot have values that are all missing.")
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:96]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+    {
+        dimension.label <- if (dimension.index == 1) "rows" else "columns"
+        paste0("The duration '", by,
+               "' cannot be applied as the ", dimension.label,
+               " in the input data are ")
+    }
+    else
+        paste0("The duration '", by,
+               "' cannot be applied as dimension ", dimension.index,
+               " in the input data is ")
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:208]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R (hunk 1 above)
Comment (JustinCCYap) [reply R/firstlast.R:247]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        stop("The input 'keep' needs to be an integer scalar or vector ",
+             "containing the number of entries to keep corresponding to ",
+             "the dimensions of the input data.")
+
+    # Check lengths of x and 'keep'
+    if ((is.null(dim.x) && length(keep) > 1))
+        stop("The input 'keep' is a vector with more than one value. It needs to ",
+             "be scalar when 'x' is non-dimensioned, e.g., when it is a ",
+             "vector.")
+    if ((is.null(dim.x) && length(keep) > 1) ||
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:89]: Fixed in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+invalidSingleValueBy <- function(by)
+{
+    !(by %in% allowed.for.by)
+}
+
+# Check that non-missing values in 'keep' and 'by' coincide, and that 'by'
+# contains permitted values
+invalidMultiValueBy <- function(by, keep)
+{
+    !setequal(which(!is.na(keep)), which(!is.na(by))) ||
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:135]: Done in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+{
+    if (any(by %in% "element"))
+    {
+        is.time.units <- by %in% allowed.time.units
+        keep.element <- keep
+        keep.element[is.time.units] <- NA
+        x <- if (is.first)
+            First(x, keep.element, "element", ...)
+        else
+            Last(x, keep.element,"element", ...)
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:164]: Fixed in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+# corresponds to the earliest period and vice versa. The parameter 'from.start'
+# indicates whether to count periods from the first date as opposed to the last
+# date when calendar is FALSE.
+periodIntegers <- function(date.times, by, calendar, from.start)
+{
+    if (calendar)
+    {
+        start.date <- structure(0, class = c("POSIXct", "POSIXt")) # unix epoch
+        if (by == "week")
+            start.date <- start.date + 3 * 24 * 60 *60 # add 3 days to unix epoch
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:228]: Added comment in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+    # Check lengths of x and 'keep'
+    if ((is.null(dim.x) && length(keep) > 1))
+        stop("The input 'keep' is a vector with more than one value. It needs to ",
+             "be scalar when 'x' is non-dimensioned, e.g., when it is a ",
+             "vector.")
+    if ((is.null(dim.x) && length(keep) > 1) ||
+        (!is.null(dim.x) && length(keep) > length(dim.x)))
+        stop("The input 'keep' is a vector with length greater than the number of ",
+             "dimensions of 'x'. Its length needs to be less than or equal to ",
+             "the number of dimensions of 'x'.")
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:93]: It is not recycled, which is consistent with `head` and `tail`

[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:93]: The dimensions which are omitted in `keep` are ignored, i.e. all elements in those dimensions are kept.

File: R/firstlast.R [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+    pattern <- paste0("^[[:blank:]]*[[:digit:]]*[[:blank:]]*(",
+                      paste0(allowed.time.units, collapse = "|"),
+                      ")s*[[:blank:]]*$")
+    is.character(x) && all(is.na(x) | grepl(pattern, x, ignore.case = TRUE))
+}
+
+firstLastByPeriod <- function(x, keep, by, calendar, is.first, ...)
+{
+    if (any(by %in% "element"))
+    {
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:157]: Added comment in 08281c682f471a7ffa02a710253dd431017d29dc

File: R/firstlast.R [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+#' @param keep An \code{integer} of the number of first elements to keep, or
+#'   if \code{by} is a unit of time, the number initial time periods to
+#'   consider when retaining elements.
+#' @param by A string that is either \code{"element"} or a unit of time:
+#'   \code{"year"}, \code{"quarter"}, \code{"month"}, \code{"week"},
+#'   \code{"day"}, \code{"hour"}, \code{"minute"}, \code{"second"}.
+#' @param calendar Whether to consider calendar periods, e.g., with calendar months,
+#'   actual months are counted, i.e., January, February etc. Otherwise months are
+#'   counted as time periods of 28 to 31 days (based on the calendar months
+#'   between start and end dates).
```
[the remark this answers is not in the document]
Comment (JustinCCYap) [reply R/firstlast.R:16]: Updated documentation in a4804b296f1307a5f822e29d29a2e64c8bbd66a1

**Comment by JustinCCYap** [comment]
A test for QTables was added in 08281c682f471a7ffa02a710253dd431017d29dc

**Pull request merged**